lib/SILOptimizer/Mandatory/MandatoryCombine.cpp

```diff
@@ -0,0 +1,223 @@
+///  instruction combines to be performed at the begining of both the Onone and
+///  also the performance pass pipelines, after the diagnostics passes have been
+///  run.  It is intended to be run before and to be independent of other
+///  transforms.
+///
+///  The intention of this pass is to be a place for mandatory peepholes that
+///  are not needed for diagnostics. Please put any such peepholes here instead
+///  of in the diagnostic passes.
+///
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-mandatory-combiner"
+#include "swift/Basic/LLVM.h"
+#include "swift/Basic/STLExtras.h"
+#include "swift/SIL/SILInstructionWorklist.h"
+#include "swift/SIL/SILVisitor.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Local.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+
+using namespace swift;
+
+namespace {
+
+class MandatoryCombiner final
+    : public SILInstructionVisitor<MandatoryCombiner, SILInstruction *> {
+
+  using Worklist = SmallSILInstructionWorklist<256>;
+
+  /// The list of instructions remaining to visit, perhaps to combine.
+  Worklist worklist;
+
+  /// Whether any changes have been made.
+  bool madeChange;
+
+  /// The number of times that the worklist has been processed.
+  unsigned iteration;
+
+  InstModCallbacks instModCallbacks;
+  SmallVectorImpl<SILInstruction *> &createdInstructions;
+  SmallVector<SILInstruction *, 16> instructionsPendingDeletion;
+
+public:
+  MandatoryCombiner(
+      SmallVectorImpl<SILInstruction *> &createdInstructions)
+      : worklist("MC"), madeChange(false), iteration(0),
+        instModCallbacks(
+            [&](SILInstruction *instruction) {
+              worklist.erase(instruction);
+              instructionsPendingDeletion.push_back(instruction);
+            },
+            [&](SILInstruction *instruction) { worklist.add(instruction); }),
+        createdInstructions(createdInstructions){};
+
+  /// Base visitor that does not do anything.
+  SILInstruction *visitSILInstruction(SILInstruction *) { return nullptr; }
+
+  void addReachableCodeToWorklist(SILFunction &function) {
+    SmallBlotSetVector<SILBasicBlock *, 32> blockWorklist;
+    SmallBlotSetVector<SILBasicBlock *, 32> blocksVisited;
+    SmallVector<SILInstruction *, 128> instructions;
+
+    blockWorklist.insert(&*function.begin());
+    while (!blockWorklist.empty()) {
+      auto *block = blockWorklist.pop_back_val().getValueOr(nullptr);
+      if (block == nullptr) {
+        continue;
+      }
+
+      if (!blocksVisited.insert(block).second) {
+        continue;
+      }
+
+      for (auto iterator = block->begin(), end = block->end(); iterator != end;) {
+        auto *instruction = &*iterator;
+        instructions.push_back(instruction);
```

Comment:
Question. Is this where there would be the check if we want to visit this or not?

Reply:
Checking whether an instruction is trivially dead if so not visiting it, you mean?

Reply:
Yes. TBH I think it does make sense here to eliminate trivially dead instructions. Feel free to do that in a subsequent commit or here. A quick thoughts (again this is just off the top of my head): rather than just not adding them (and deleting them), why not maintain a separate local list as well and then pass them to recursivelyDeleteTriviallyDeadInstructions?

Reply:
Okay, it's now checking for trivially dead, adding those to a list (SmalLVector) of instructions to delete, and then passing that list to recursivelyDeleteTriviallyDeadInstructions.  Is that what you had in mind?

Reply:
yes.